pkg/api/kptfile/v1alpha2/types.go

```diff
@@ -0,0 +1,201 @@
+
+	// Pipeline declares the pipeline of functions.
+	Pipeline *Pipeline `yaml:"pipeline,omitempty"`
+
+	// Inventory contains parameters for the inventory object used in apply.
+	Inventory *Inventory `yaml:"inventory,omitempty"`
+}
+
+// OriginType defines the type of origin for a package
+type OriginType string
+
+const (
+	// GitOrigin specifies a package as having been cloned from a git repository
+	GitOrigin OriginType = "git"
+)
+
+// UpdateStrategyType defines the strategy for updating a package from upstream.
+type UpdateStrategyType string
+
+const (
+	// ResourceMerge performs a structural schema-aware comparison and
+	// merges the changes into the local package.
+	ResourceMerge UpdateStrategyType = "resource-merge"
+	// FastForward fails without updating if the local package was modified
+	// since it was fetched.
+	FastForward UpdateStrategyType = "fast-forward"
+	// ForceDeleteReplace wipes all local changes to the package.
+	ForceDeleteReplace UpdateStrategyType = "force-delete-replace"
+)
+
+// Upstream is a reference to where the package is fetched from.
+type Upstream struct {
+	// Type is the type of origin.
+	Type OriginType `yaml:"type,omitempty"`
+
+	// Git contains information on the origin of packages fetched from a git repository.
+	Git Git `yaml:"git,omitempty"`
+
+	// UpdateStrategy defines how a package is updated from upstream.
+	UpdateStrategy UpdateStrategyType `yaml:"updateStrategy,omitempty"`
+}
+
+// PackageMeta contains metadata such as license, documentation, etc.
+// These fields are not used for any functionality in kpt and are simply passed through.
+type PackageMeta struct {
+	// URL is the location of the package.  e.g. https://github.com/example/com
+	URL string `yaml:"url,omitempty"`
+
+	// Email is the email of the package maintainer
+	Email string `yaml:"email,omitempty"`
+
+	// License is the package license
+	License string `yaml:"license,omitempty"`
+
+	// Version is a logical package version (ignored by kpt)
+	Version string `yaml:"version,omitempty"`
+
+	// Tags enables humans and tools to attach arbitrary package metadata.
+	Tags []string `yaml:"tags,omitempty"`
+
+	// Man is the path to documentation about the package
+	Man string `yaml:"man,omitempty"`
+
+	// ShortDescription contains a short description of the package.
+	ShortDescription string `yaml:"shortDescription,omitempty"`
+}
+
+// Subpackages declares a local or remote subpackage.
+type Subpackage struct {
+	// Name of the immediate subdirectory relative to this Kptfile where the suppackage
+	// either exists (local subpackages) or will be fetched to (remote subpckages).
+	// This must be unique across all subpckages of a package.
+	LocalDir string `yaml:"localDir,omitempty"`
+	// Whether a subpackage is local or remote is determined by whether Upstream is specified.
+	// Upstream is a reference to where the package is fetched from.
+	Upstream Upstream `yaml:"upstream,omitempty"`
```

Comment:
We either need a `Type` field here, or `Upstream` needs to be a pointer. We don't want to have to look at fields within the `Upstream` struct.

Reply:
Yep, making this a pointer makes sense. In fact most fields in Kptfile should be pointers as they're optional.